text/0095-cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,567 @@
+});
+```
+
+> These are the defaults that will be part of tsdoc, and not part of the README -
+> * allowUnauthenticated: false
+> * authenticationFlow: `AuthenticationFlow.ENHANCED`
+
+### Identity Providers
+
+Identity pools can be configured with external providers for sign in. These can be from facebook, google, twitter as
+well as, "login with Amazon" and Cognito user pools. Use the `identityProviers` property to configure these providers.
+Learn more about external identity providers at [Cognito's
+documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/external-identity-providers.html). The CDK
+documentation of the `identityProviders` property has more details on how to configure an identity pool with each of
+these external providers.
+
+> Internal Note: IdentityPool-SupportedLoginProviders takes a JSON with all of the providers configured.
+
+Besides the external identity providers, Cognito identity pools also support developer authenticated identities. Read
+[the documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/developer-authenticated-identities.html)
+to learn more about developer authenticated identities. This can be configured via the property `developerProvider`.
+
+Code snippet with all of the external providers configured -
+
+```ts
+new IdentityPool(this, 'myidentitypool', {
+  // ...
+  identityProviders: {
+    amazon: { appId: 'amzn1.application.188a56d827a7d6555a8b67a5d' },
+    facebook: { appId: '7346241598935555' },
+    google: { clientId: '123456789012.apps.googleusercontent.com' },
+    twitter: { consumerKey: 'xvz1evFS4wEEPTGEFPHBog', consumerSecret: 'kAcSOqF21Fu85e7zjz7ZN2U4ZRhfV3WpwPAoE3Z7kBw' },
+    developerProvider: 'login.mycompany.myapp'
+  }
+});
+```
+
+### Access Control
+
+Identity pools are configured with IAM roles from which temporary credentials will be provided to users of this pool
+when they need to access AWS resources. The IAM roles are configured with permissions based on what the users are
+allowed to access.
+
+An identity pool can have separate IAM roles configured for its authenticated users (authenticated via one of the
+identity providers as above) and unauthenticated users. They can be specified via the properties `authenticatedRole` and
+`unauthenticatedRole` properties. When no role is specified against these properties, CDK will create a new role with
+the permissions to Amazon Mobile Analytics and Amazon Cognito Sync. This is similar to the behaviour of the AWS console.
+Read the [IAM Roles documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/iam-roles.html) to learn
+more on how to set up your own IAM roles.
+
+```ts
+const authenRole = new iam.Role(this, 'cognito-authenticated-role', { ... });
+
+const unAuthenRole = new iam.Role(this, 'cognito-unauthenticated-role', { ... });
+
+new IdentityPool(this, 'my-awesome-id-pool', {
+  // ...
+  // ...
+  authenticatedRole: authenRole,
```

Comment:
Does it make sense to autocreate these?

```ts
const pool = new IdentityPool(this, 'my-awesome-id-pool');

bucket.grantRead(pool.unauthenticatedRole);
bucket.grantReadWrite(pool.authenticatedRole);
```

Especially since I believe the `assumedBy` parameter for the roles is going to be quite tricky in practice, no?

Reply:
> Does it make sense to autocreate these?

Yes, there should be a default role created for these with the `authenticatedRole` property to specify an existing role. I seem to have missed this and will add that in.

> Especially since I believe the assumedBy parameter for the roles is going to be quite tricky in practice, no?

The doc states -- "The CDK, by default, adds a trust policy with these configured to all roles used by the IdentityPool. To disable this, set the configureRoleTrust to false."

Essentially I'm hoping to call `role.assumeRolePolicy.addStatements(...)` and add the correct trust parameters.

Does this satisfy?